Local database upgrade to NeDb

lusbenjamin opened the pull request:
This PR moves our local persistence over to [nedb](https://github.com/louischatriot/nedb).

### Motivation
Long-time users of our tool have started to build up a non-trivial amount of historical data (25+MB). This currently causes some performance issues with our current approach to local persistence, which is to put all of it together in one giant [electron-store](https://github.com/sindresorhus/electron-store). Although some performance issues may also be related to IPC traffic volume, we currently believe most underlying issues are due to slow `electron-store` reads and writes when the player data file grows too large in size.

### NOTE: `npm install` and backup your local data file first!
`npm install` and backup your local data file first!

### nedb vs electron-store
pros:
  - faster
  - enforces asynchronous use
  - can handle more data
  - more powerful querying API
  - similar to our server persistence tier (MongoDb)

cons:
  - pain-in-the-🍑 to use the `callback`-style API
  - requires us to store well-formed documents instead of individual fields
  - files-on-disk are less friendly to external management since they aren't just JSON

### Approach
- Deliberately _leaves the database schema alone_ to simplify migration (we can refactor the schema later)
- creates a new `LocalDatabase` interface to isolate and encapsulate our current local persistence calls and enforce asynchronous use (i.e. pain-in-the-🍑 callbacks galore). The API design is roughly "half-way between" `electron-store` and `nedb`.
- refactors our existing code to use `LocalDatabase` instead of raw `electron-store`
  - `main` process depends on application store to set up auto-updater
  - `background` process depends on application store and personal-data store
- wraps our old `electron-store` dependency to implement `ElectronStoreDatabase` as an updated `LocalDatabase` storage option.
  - there is hacky logic to combine together `settings.json` and `remember.json` into a single "application" store
- imports `nedb` to the project and wraps it to implement `NeDbDatabase`, which meets our project in the middle as an faster and more scalable `LocalDatabase` storage option.
  - there is hacky logic to wrap "bare" values in proper documents
- adds fancy migration logic (see next section)

#### Migration Logic
This PR does 2 things automatically:
1. A one-time, bulk migration of player data from `electron-store` to `nedb` during their first boot of the tool. The old JSON file is left in place, the new `nedb` becomes "truth".
2. Creates a new `nedb` application database and continuously synchronizes it with the old `electron-store`. The old JSON file is left in place and remains "truth".

Since this PR has interfaces for both types of database, we have a fair amount of flexibility in the short term to handle data portability, migration, and backwards compatibility. (As soon as we start to refactor our schema, this will become much more difficult.)

##### Details about the bulk migration
- During tool initialization, make both a legacy (electron-store) and a current (nedb) player db
- attempt to load the `cards` data from both sources, then:
  - iff the legacy db has data and the current db has no data, it will do a full data migration
  - in all other cases, just launch assuming the current db is up-to-date

##### Details about the application database
Using a `MigrationDatabase` lets us seamlessly boot from program state on either side of the migration without forcing a user login or losing "brittle" application-level settings. Once we have deployed this code everywhere, we can consider moving the application store "truth" to `nedb` and sunset `electron-store` entirely.

### BONUS Backport Data to JSON
This PR also adds a new "Backport Data to Legacy JSON" button to the "Settings >> Data >> Local Data" section:
<img width="983" alt="Annotation 2019-11-10 020308" src="https://user-images.githubusercontent.com/14894693/68551567-168bfb00-03c3-11ea-86c4-28e896b11647.png">
- This should help with debugging and user support! (JSON is just more portable)
- The new backport button helps us support folks like @AnnanFay (at least for now)
- Since we still control the JSON-file resource location, this gives us a manual backwards migration path for now (so you can hit the button and then easily jump back to `master` or `develop`)

### Future TODOs
- After we deploy this to stable "for a while", create another PR to sunset `electron-store`
- Consider a longer-term JSON export feature (not necessarily backwards compatible)

Manuel-777 commented:
Im curious as of how could we "detect" when to migrate. The migrator works, as far as I tested.

The flow I imagined;
- Check if local store exists
  - If it does use `MigrationDb`
    - Rename old store file to avoid loading it in the future _(we are nice people and nice people do not delete other's files)_
  - else use `NeDb`

Also we SERIOUSLY need a way to know if the migration is taking place and how much it loaded. What happens if we close tool midway?

lusbenjamin commented:
This now has a proposed set of migration logic (please comment!) and a relatively safe way to convert user data. This still needs a fair amount of testing, but it is definitely ready for review.

lusbenjamin commented:
### Migration Demo
![migration-demo](https://user-images.githubusercontent.com/14894693/68551874-e134dc80-03c5-11ea-82c7-7e4c15efff38.gif)
@Manuel-777 I put some work into solving some of your earlier points:
- console log gives detail about migration logic and callback flow for devs
- dialog messages give users a heads up (wording may need help)
- nothing happens during the migration and it is clear when it completes
- simplified migration logic check so that it is based purely on the existence of the databases
- added a new area to the settings page with a backport feature for @AnnanFay

lusbenjamin commented on src/window_background/background.js:
```diff
@@ -51,110 +55,107 @@ globals.toolVersion = (app || remote.app)
+  appDb.find("", "logUri", (err, logUri) => {
+    appDb.find("", "email", (err, email) => {
+      appDb.find("", "token", (err, token) => {
+        appDb.find("", "settings", (err, appSettings) => {
```
I hate this. 😡

Manuel-777 replied:
It doesnt get much better with promises?

`appDb.find("", "logUri").then(logUri =>
  appDb.find("", "email").then(email =>
    appDb.find("", "token").then(token =>
      appDb.find("", "settings").then(appSettings =>
        loadAppSettings(logUri, email, token, appSettings)
      )
    )
  )
);`

Manuel-777 replied:
Things like this can be simplified using Mongo queries, using a single find() and a projection. But that is not backwards compatible with electron-store.

lusbenjamin commented:
Does anyone have experience with Promises? I began this PR long, long, ago in the pre-TypeScript era, but I think there may be some opportunities in the post-TypeScript world to make and break some Promises?

Manuel-777 commented:
> Does anyone have experience with Promises? I began this PR long, long, ago in the pre-TypeScript era, but I think there may be some opportunities in the post-TypeScript world to make and break some Promises?

I used them in the metadata script. Its not that hard once you get it :)

lusbenjamin commented:
Hmmm. You're right, we already have a bunch of async code and some Promises. I have no idea what I'm talking about, Promises were already in JS standard. I'm looking for a better pattern than the `callback`-confetti bomb that this PR drops on our repo. 🎊 🎊 🎊

lusbenjamin commented:
If you are taking this PR for a spin, the "Backport Data to Legacy JSON" button also makes it easy to quickly "bring your data" back to the main `develop` branch without rescanning your output log. Convenient... if you trust that this PR got it all correct in the first place. ;)

Thaoden commented:
Would it be possible to save the "new" version of the local database to a different folder? I _think_ users can then continue using the old and the new version side by side and the impact of a broken / failed migration is lower.

lusbenjamin commented:
> Would it be possible to save the "new" version of the local database to a different folder?
Most definitely. We could create a full "Data Export" feature with a native file browser save dialog without much additional effort. The reason this button writes to exactly the same resource location as current `master` would is _because_ I was trying to allow users to continue using the old and the new version side by side. For example, I believe if you diff your personal data JSON before and after doing an NeDb export (just diff the current file to the backup copy that you totally remembered to make), `electron-store` will make a minimum amount of changes and leave most of your prior data in-place. (This was how I did a bunch of empirical testing.)

Regardless of this potential migration, this PR demonstrates that it would not be hard to add an "Export All Yer Data" feature at some point.

lusbenjamin commented:
@Thaoden made the excellent point that given the breaking nature of this build, it might be best aimed at a v3.0 release.

Manuel-777 commented:
> @Thaoden made the excellent point that given the breaking nature of this build, it might be best aimed at a v3.0 release.

I was also going to suggest that ! Syncccc

Thaoden commented on src/shared/local-database.ts:
```diff
@@ -313,24 +386,31 @@ class ElectronStoreDb {
+  appStore?: Datastore;
+  playerStore?: Datastore;
+
+  constructor(useBulkFirstpass?: boolean) {
```
Instead of using a `boolean?`, you can pass in a default value: `useBulkFirstPass: boolean = false`.

Thaoden commented on src/shared/local-database.ts:
```diff
@@ -313,24 +386,31 @@ class ElectronStoreDb {
+    this.dbName = "";
+    // whether or not to imitate old electron-store bulk-write
+    // behavior during the first pass of the log (deprecated workaround)
+    this.useBulkFirstpass = useBulkFirstpass || false;
```
Then you don't need this check.

Thaoden commented on src/shared/local-database.ts:
```diff
@@ -193,7 +193,7 @@ const nonDocFields = [
 
```
(The position of this comment does not have any meaning)
This file is quite long and contains at least 4 different classes. Maybe separate each class in its own file.

Thaoden commented on src/shared/LocalDatabase.ts:
```diff
@@ -0,0 +1,96 @@
+import { ElectronStoreDatabase } from "./ElectronStoreDatabase";
+import { NeDbDatabase } from "./NeDbDatabase";
+
+class MigrationDatabase implements LocalDatabase {
```
nit: This file should probably be named `MigrationDatabase.ts` (`LocalDatabase` is apparently imported)

Thaoden commented on src/shared/databaseUtil.ts:
```diff
@@ -0,0 +1,241 @@
+  }
+}
+
+export interface LocalDatabase {
```
I guess this interface is big enough to be separated in its own file, too :)

Thaoden replied:
Maybe make this a class and move the `defaultCallback` function into here, too.

Thaoden commented on src/shared/NeDbDatabase.ts:
```diff
@@ -0,0 +1,220 @@
+    return this.dbName === "application" ? this.appStore : this.playerStore;
+  }
+
+  init(dbName: string, arenaName: string) {
```
You can move the `init` function call into the constructor and add the necessary arguments to the constructor signature. That way, you can be sure that properties like `this.dbName` are always set and you don't need the `if(!this.datastore)` checks anymore.
As an aside, you can then mark this function as well as `this.dbName` as `private`.

lusbenjamin replied:
Our `LocalDatabase`s preserve the distinction in database lifecycle between instantiation and initialization. Both electron-store and nedb have various reasons that you would want to have a handle on an instance but not yet want to spin up the machinery or touch the disk yet.

lusbenjamin replied:
...and now that I re-read the implementations, I see that we don't actually even instantiate until we initialize. 🤷‍♀ May as well just combine them if we don't need the lifecycle separate?

Thaoden replied:
I don't know enough of NeDb. But if you can always assume you have a valid datastore, I guess it saves you some gray hairs.
That aside, what was the use case? As soon as the tool starts up, it reads settings from the datastore, so it needs an initialised db connection.

lusbenjamin replied:
We keep user data and application level data separate and instantiate the databases at different times. This is (possibly inaccurate) simplification:
1. boot up the app database and load the app boot settings
2. boot the various processes like the main display and background log reader
3. scan the log to figure out who the user is
4. use that user information to initialize the user database
5. re-scan the log to process data for that user and keep booting...

In our idealized future, our tool would support N user-level databases, swapping between them on the fly whenever it detected that the Arena user had changed in the log.

lusbenjamin commented on src/shared/db/NeDbDatabase.ts:
```diff
@@ -0,0 +1,241 @@
+        callback(null, total);
+      }
+    };
+    recursiveHelper(allData, 0, this.upsert);
```
Open design bounty: what is a better way to implement `upsertAll`? Can we do something more efficient in time or space? We have to upsert destructively, so we can't call the standard `nedb` bulk insert. We also want to preserve the order of entries if possible, though I'm not sure that has any meaning for how nedb stores data?

mikespo23 commented:
Is it possible to create a Json or API request from this software or MTG Arena to speak with a Wordpress Plugin or something similar?

lusbenjamin commented:
Welcome! The best way to discuss building a potential new feature or API would be to file a new github enhancement with more detail about what you are trying to do and why. If you are curious about the existing server API, the best thing to do would probably be to join the Discord and ask there.

mikespo23 commented:
Ben Thanks

Essentially this would be a wordpress plug-in that would update decks that
are doing well, rankings and reports from Arena.

Below is my Discord ID could you send me a request to join the channel as
well as the link to where to post more detailed information I really
appreciate it thank you!!

Discord ID- Mike3#7214

On Sun, Nov 17, 2019 at 2:55 PM Benjamin Lu <notifications@github.com>
wrote:

> Welcome! The best way to discuss building a potential new feature or API
> would be to file a new github enhancement with more detail about what you
> are trying to do and why. If you are curious about the existing server API,
> the best thing to do would probably be to join the Discord and ask there.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Manuel-777/MTG-Arena-Tool/pull/656?email_source=notifications&email_token=ANM4BVLURBE3U5LPSLZZ6QLQUGOUJA5CNFSM4JEYNXVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEIULMQ#issuecomment-554780082>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANM4BVNLTTW7XLYFB5XFIK3QUGOUJANCNFSM4JEYNXVA>
> .
>
-- 
Best Regards,
Michael Esposito
410-934-2063
443-891-3840

IT Specialist
Linkedin <https://www.linkedin.com/in/michael-esposito-81860191/>

lusbenjamin commented:
We appreciate all bug reports and feature suggestions: https://discord.gg/K9bPkJy

lusbenjamin commented:
now stacked on #754
